### ISSUE TITLE
Orleans 7 Object not set Error

### ISSUE THREAD

**Issue by sjbthfc2** [dotnet/orleans#8563]
Hi,

We are trying to upgrade one of our services to Orleans 7.2.1 from 3.x and we have an issue both running locally and when our silo and client are deployed. The error occurs when the client tried to make a call to a grain (Stateless Worker) which seems to blow up with the following error:

`
Object reference not set to an instance of an object.
`

originating from:

`   at Orleans.OutsideRuntimeClient.SendRequestMessage(GrainReference target, Message message, IResponseCompletionSource context, InvokeMethodOptions options)
   at Orleans.OutsideRuntimeClient.SendRequest(GrainReference target, IInvokable request, IResponseCompletionSource context, InvokeMethodOptions options)
   at Orleans.Runtime.GrainReferenceRuntime.<.ctor>b__5_0(GrainReference reference, IResponseCompletionSource callback, IInvokable body, InvokeMethodOptions options)
   at Orleans.Runtime.OutgoingCallInvoker``1.<Invoke>d__38.MoveNext()
   at Orleans.Runtime.OutgoingCallInvoker``1.<Invoke>d__38.MoveNext()
   at Orleans.Runtime.GrainReferenceRuntime.<InvokeMethodWithFiltersAsync>d__14.MoveNext()
   at Me.MelCloudHome.MelCloudCommunicationIngestion.Worker.IngestionEventHandlers.MelCloudUploadReceivedEventHandler.<Handle>d__3.MoveNext() in C:\Users\browns\source\repos\MELCloudHome\services\CommunicationServices\MelCloudCommunicationService\src\MelCloudCommunicationIngestionWorker\IngestionEventHandlers\MelCloudUploadReceivedEventHandler.cs:line 27`

The calling code is:
`
                    var dispatchGrain = _client.GetGrain<IDispatcherGrain>(0);
                    await dispatchGrain.DispatchUpload(notification.MacAddress, notification.Xml);`

This occurs on a simple client and silo using local host clustering.

Any ideas what the issue could be? We are completely stumped.

Many thanks

**Comment by ReubenBond** [comment]
This most likely means that you have not connected to the cluster yet and are using the client before connection has completed.

If you show your startup code, we can show you how to update it to fix it.
We should make the error message here more obvious...

**Comment by ReubenBond** [comment]
Yes, you need to ensure you call `await host.StartAsync()` before you use the client. I cannot see that part of your code here.

**Comment by sjbthfc2** [comment]
Hi,

We do already have `await host.StartAsync()` in the client and it doesn't work. I have omitted service and logging configuration in the code below but the client host looks like:

```
IHost host = Host.CreateDefaultBuilder(args)
    .ConfigureAppConfiguration((hostContext, builder) =>
    {
            client
                .UseLocalhostClustering(30001, "MelCloudCommunicationService", "CommunicationServices")
                .AddOutgoingGrainCallFilter(context =>
                {
                    RequestContext.Set("IsExceptionConversionEnabled", true);
                    return context.Invoke();
                })
                .Configure<ClientMessagingOptions>(options =>
                {
                    options.ResponseTimeout = TimeSpan.FromMinutes(10);
                })
                .UseConnectionRetryFilter<ClientConnectRetryFilter>();
     })
    .UseConsoleLifetime()
    .Build();

await host.StartAsync();
```

**Comment by sjbthfc2** [comment]
Just to add, this error still happens if I spin up the silo and then later spin up the client. The error occurs when calling a grain methid after everything spins up. This is the grain call, IDispatcherGrain is a one way method on a stateless worker:

`var dispatchGrain = _client.GetGrain<IDispatcherGrain>(0);
                    await dispatchGrain.DispatchUpload(notification.MacAddress, notification.Xml);`

**Comment by sjbthfc2** [comment]
Hi,

We have finally fixed the issue, we rewrote the client config for 7 from scratch using the docs example rather than tweaking what we had with 3.x and it is working now. I guess we were thrown off by the fact the the client did intermittently connect and worked calling a grain, and then other times threw the object not set error rather than simply the client not connecting.

Thanks again for your help though.
